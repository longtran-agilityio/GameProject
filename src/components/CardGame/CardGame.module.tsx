import { styled, Box, Typography } from '@mui/material'

export const StyledCardGame = styled(Box)(({ theme }) => ({
  display: 'flex',
  flexDirection: 'column',
  borderRadius: theme.spacing(2),
  overflow: 'hidden',
  backgroundColor: theme.palette.background.paper,
  width: theme.spacing(30),
}))

export const StyledInfoCard = styled(Box)(({ theme }) => ({
  display: 'flex',
  flexDirection: 'column',
  padding: theme.spacing(1.5, 1.5, 3),
  color: theme.palette.text.secondary,
}))

export const StyledImage = styled('img')(({ theme }) => ({
  width: '100%',
  height: '100%',
}))

export const StyledBox = styled(Box)(({ theme }) => ({
  display: 'flex',
  justifyContent: 'space-between',
}))

export const StyledTitle = styled(Typography)(({ theme }) => ({
  color: theme.palette.text.primary,
  fontWeight: 900,
  fontSize: theme.spacing(1.8),
}))

export const StyledReleased = styled(Typography)(({ theme }) => ({
  fontSize: theme.spacing(1.3),
}))

export const StyledGenres = styled(Typography)(({ theme }) => ({
  fontSize: theme.spacing(1.3),
}))
