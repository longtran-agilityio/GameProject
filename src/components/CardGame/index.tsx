import AddIcon from '@mui/icons-material/Add'
import DoneIcon from '@mui/icons-material/Done'
import { Link, generatePath } from 'react-router-dom'
import Skeleton from 'react-loading-skeleton'
// component
import { StyledButton } from '@webapp/components/Button/Button.module'

// styles
import { StyledPrice } from '@webapp/components/ShoppingCart/ShoppingCart.module'
import {
  StyledCardGame,
  StyledImage,
  StyledBox,
  StyledTitle,
  StyledReleased,
  StyledGenres,
  StyledInfoCard,
} from './CardGame.module'

import { PageUrls } from '@webapp/constants/pageUrl'
import { IGame, IAddGame } from '@webapp/interfaces/game'
import { useCart } from '@webapp/contexts/games/cartProvider'
import { useAuth } from '@webapp/contexts/useAuth'
import { Box } from '@mui/material'
import { Suspense } from 'react'

interface ICardGame {
  game: IGame
  addGame: ({ id, cardGame }: IAddGame) => void
}
const CardGame = ({ game, addGame }: ICardGame) => {
  const { cartsList } = useCart()
  const handleAddToCart = () => {
    addGame({ id: game.id, cardGame: game })
  }
  const userAuthentication = useAuth()

  const addGameBtnAuthentication = userAuthentication ? (
    <StyledButton
      onClick={handleAddToCart}
      size='medium'
      endIcon={<AddIcon />}
      disableRipple
      sx={{
        '&:hover': {
          color: '#92f',
        },
      }}
    >
      Add to cart
    </StyledButton>
  ) : (
    <Link to={PageUrls.LOGIN}>
      <StyledButton size='medium' endIcon={<AddIcon />} disableRipple>
        Add to cart
      </StyledButton>
    </Link>
  )

  return (
    <StyledCardGame>
      <Link
        to={generatePath(PageUrls.GAME_DETAIL, { id: game.id.toString() })}
        aria-label={game.name}
      >
        <Suspense fallback={<Skeleton count={3} />}>
          <Box sx={{ height: '200px' }}>
            <picture>
              <source srcSet={game.background_image} media='min-width: 300px' />
              <source srcSet={game.background_image} media='min-width: 400px' />

              <StyledImage
                srcSet={`${game.background_image} 2x`}
                sizes='(max-width: 300px) 300px, (max-width: 300'
                src={game.background_image}
                alt={game.name}
              />
            </picture>
          </Box>
        </Suspense>
      </Link>

      <StyledInfoCard>
        <StyledBox>
          {cartsList?.find((item) => item.id === game.id) ? (
            <StyledButton size='medium' endIcon={<DoneIcon />} disabled disableRipple>
              Added
            </StyledButton>
          ) : (
            addGameBtnAuthentication
          )}
          <StyledPrice variant='body1'>${game.prices}</StyledPrice>
        </StyledBox>
        <Link to={generatePath(PageUrls.GAME_DETAIL, { id: game.id.toString() })}>
          <StyledTitle>{game.name}</StyledTitle>
        </Link>
        <StyledReleased>Released: {game.released}</StyledReleased>
        <StyledGenres>Genres: {[game.genres.map(({ name }) => name)].join(',')}</StyledGenres>
      </StyledInfoCard>
    </StyledCardGame>
  )
}

export default CardGame
