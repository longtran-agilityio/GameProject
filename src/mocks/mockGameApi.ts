import { faker } from '@faker-js/faker'

export const games = [
  {
    id: faker.datatype.number(),
    name: faker.name.firstName(),
    prices: faker.datatype.number(),
    released: faker.commerce.product(),
    background_image: faker.image.cats(),
    description_raw: faker.commerce.product(),
    genres: [
      {
        name: faker.name.firstName(),
      },
    ],
    developers: [
      {
        id: faker.datatype.number(),
        name: faker.name.firstName(),
        slug: faker.name.firstName(),
      },
    ],
    platforms: [
      {
        platform: {
          id: faker.datatype.number(),
          name: faker.name.firstName(),
          slug: faker.name.firstName(),
        },
      },
    ],
    publishers: [
      {
        id: faker.datatype.number(),
        name: faker.name.firstName(),
      },
    ],
  },
]
