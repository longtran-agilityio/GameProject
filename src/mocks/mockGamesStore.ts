import { IGame } from '@webapp/interfaces/game'

export const game: IGame[] = [
  {
    id: 3498,
    name: 'Grand Theft Auto V',
    prices: 50.12,
    released: '2013-09-17',
    background_image: 'https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg',
    description_raw: '',
    genres: [
      {
        name: 'Action',
      },
      {
        name: 'Adventure',
      },
    ],
    developers: [
      {
        id: 3524,
        name: 'Rockstar North',
        slug: 'rockstar-north',
      },
      {
        id: 10,
        name: 'Rockstar Games',
        slug: 'rockstar-games',
      },
    ],
    platforms: [
      {
        platform: {
          id: 4,
          name: 'PC',
          slug: 'pc',
        },
      },
    ],
    publishers: [
      {
        id: 2155,
        name: 'Rockstar Games',
      },
    ],
  },
]
