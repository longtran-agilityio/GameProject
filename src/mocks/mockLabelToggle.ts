import { PageUrls } from '../constants/pageUrl'

export const labelToggle = [
  {
    label: 'Login',
    currentUrl: PageUrls.LOGIN,
  },
  {
    label: 'Register',
    currentUrl: PageUrls.REGISTER,
  },
]
