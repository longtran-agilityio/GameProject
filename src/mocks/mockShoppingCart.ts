export const mockShoppingCart = [
  {
    id: 1,
    name: 'Pokemon',
    price: 19.0,
  },
  {
    id: 2,
    name: 'Pokemon1',
    price: 19.0,
  },
  {
    id: 3,
    name: 'Pokemon2',
    price: 19.0,
  },
  {
    id: 4,
    name: 'Pokemon3',
    price: 19.0,
  },
  {
    id: 5,
    name: 'Pokemon4',
    price: 19.0,
  },
]
