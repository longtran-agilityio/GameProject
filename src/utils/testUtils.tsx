import { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { ThemeProvider } from '@mui/material'
import { theme } from '@webapp/themes/index'

const AllBrowserRouter = ({ children }: { children: React.ReactNode }) => (
  <BrowserRouter>
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  </BrowserRouter>
)

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, 'wrapper'>) =>
  render(ui, { wrapper: AllBrowserRouter, ...options })

export * from '@testing-library/react'
export { customRender as render }
