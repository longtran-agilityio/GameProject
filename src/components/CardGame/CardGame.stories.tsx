// libs
import { ThemeProvider } from '@mui/material'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import { BrowserRouter } from 'react-router-dom'

// component
import CardGame from '.'
import { theme } from '@webapp/themes/index'

import { game } from '@webapp/mocks/mockGamesStore'

export default {
  title: 'Components/CardGame',
  component: CardGame,
} as ComponentMeta<typeof CardGame>

const Template: ComponentStory<typeof CardGame> = (args) => (
  <BrowserRouter>
    <ThemeProvider theme={theme}>
      <CardGame {...args} />
    </ThemeProvider>
  </BrowserRouter>
)

export const CardGameComponent = Template.bind({})

CardGameComponent.args = {
  game: game[0],
}
