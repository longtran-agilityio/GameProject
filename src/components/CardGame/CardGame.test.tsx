// libs
import { render, screen } from '@webapp/utils/testUtils'

// component
import CardGame from '.'

import { game } from '@webapp/mocks/mockGamesStore'

const onClick = jest.fn()
const CardGameComponent = () => render(<CardGame game={game[0]} addGame={onClick} />)

describe('CardGame', () => {
  test('should matches snapshot', () => {
    const { container } = CardGameComponent()
    expect(container).toMatchSnapshot()
  })

  test('should render correctly', () => {
    CardGameComponent()
    expect(screen.getByText(`$${game[0].prices}`)).toBeInTheDocument()
  })
})
