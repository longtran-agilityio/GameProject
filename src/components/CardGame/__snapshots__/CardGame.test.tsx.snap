// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardGame should matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-sq7b5v"
  >
    <a
      href="/game-detail/3498"
    >
      <img
        class="css-1kw5lk5"
        src="https://media.rawg.io/media/games/456/456dea5e1c7e3cd07060c14e96612001.jpg"
      />
    </a>
    <div
      class="MuiBox-root css-1gx70ut"
    >
      <div
        class="MuiBox-root css-mbiude"
      >
        <a
          href="/login"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-17rxn03-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Add to cart
            <span
              class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AddIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
          </button>
        </a>
        <p
          class="MuiTypography-root MuiTypography-body1 css-wfopbl-MuiTypography-root"
        >
          $
          50.12
        </p>
      </div>
      <a
        href="/game-detail/3498"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-1hvo8i1-MuiTypography-root"
        >
          Grand Theft Auto V
        </h6>
      </a>
      <p
        class="MuiTypography-root MuiTypography-body1 css-10rgspy-MuiTypography-root"
      >
        Released: 
        2013-09-17
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-10rgspy-MuiTypography-root"
      >
        Genres: 
        Action,Adventure
      </p>
    </div>
  </div>
</div>
`;
